src/classes/UTIL_CurrencyCache.cls

```diff
@@ -0,0 +1,310 @@
+                = new Map<String, UTIL_CurrencyCache.CurrencyData>();
+
+        // Even if ACM is enabled, the Admin may choose to not use dated conversion rates for roll-ups.
+        Boolean useDatedConversionRates = UTIL_Currency.getInstance().isAdvancedCurrencyManagementEnabled()
+                && householdSettings.Use_Dated_Conversion_Rates__c == true;
+
+        if (useDatedConversionRates) {
+            // When ACM is enabled, query the DatedConversionRate table to retrieve the effective dates
+            // but only for the specific currency code requested.
+            String soql = 'SELECT IsoCode, ConversionRate, StartDate ' +
+                    'FROM DatedConversionRate ' +
+                    'WHERE IsoCode IN :currCodes ' +
+                    'ORDER By IsoCode, StartDate DESC';
+            for (SObject dcr : database.query(soql)) {
+
+                String currCode = (String) dcr.get('IsoCode');
+                UTIL_CurrencyCache.CurrencyData currData = mapOfCurrencyDataByCurrencyCode.get(currCode);
+                if (currData == null) {
+                    currData = new UTIL_CurrencyCache.CurrencyData();
+                }
+
+                Date dt = (Date) dcr.get('StartDate');
+                Decimal rt = (Decimal) dcr.get('ConversionRate');
+                currData.effectiveDates.add(dt);
+                currData.rates.add(rt);
+                mapOfCurrencyDataByCurrencyCode.put(currCode, currData);
+            }
+
+
+        }
+        if (UserInfo.isMultiCurrencyOrganization()) {
+            // If ACM is not enabled, query the standard rate table and create a single rate
+            // instance per currency code for all currency codes with an effective date of 1/1/900.
+            String soql = 'SELECT IsoCode, ConversionRate, IsCorporate, DecimalPlaces FROM CurrencyType WHERE IsActive = True';
+            if (useDatedConversionRates) {
+                // If dated exchange rates are in affect, we still need to capture the 'default'
+                // standard rate to use in the specific scenario of determing the exchange rate
+                // against the legacy rollups that are based on aggragate soql.
+                soql += ' AND IsoCode IN :currCodes';
+            }
+            for (SObject scr : database.query(soql)) {
+                String currCode = (String) scr.get('IsoCode');
+
+                UTIL_CurrencyCache.CurrencyData currData = mapOfCurrencyDataByCurrencyCode.get(currCode);
+                if (currData == null) {
+                    currData = new UTIL_CurrencyCache.CurrencyData();
+                }
+
+                currData.defaultRate = (Decimal)scr.get('ConversionRate');
+                currData.isOrgDefault = (Boolean)scr.get('IsCorporate');
+                currData.decimalPlaces = (Integer)scr.get('DecimalPlaces');
+                mapOfCurrencyDataByCurrencyCode.put(currCode, currData);
```

Comment:
Couldn't this be overwriting the dated conversion rate data for the same code from the previous block? Can't orgs have both?

Reply:
Line 255 attempts to retrieve the CurrencyData object from the map. If that returns null, it creates a new object. Line 263 puts the new/updated object back into the map.